src/MinimiseProgramTransformer.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2018, The Souffle Developers. All rights reserved.
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file MinimiseProgramTransformer.cpp
+ *
+ * Define classes and functionality related to program minimisation.
+ *
+ ***********************************************************************/
+
+#include "AstProgram.h"
+#include "AstTransforms.h"
+#include "AstVisitor.h"
+#include <stack>
+
+namespace souffle {
+
+/**
+ * Extract valid permutations from a given permutation matrix of valid moves.
+ */
+std::vector<std::vector<unsigned int>> extractPermutations(
```

Comment:
You could put the helper methods into an anonymous namespace, since they're only used in this file.